src/harness/env/environment_container.rs

```diff
@@ -0,0 +1,516 @@
+
+        // combine them, sets self.environment_list
+        self.try_edit_values(&to_append, EditMode::Append)
+    }
+
+    /// Remove a variable from all Environments.
+    ///
+    /// Removes either a whole variable with all its values, or some values of a variable,
+    /// depending on the content of `to_remove`. For example:
+    /// ```notest
+    /// env_container.remove_from_environments([vec!["FOO", "1", "2"], vec!["BAR"]])
+    /// // removes any mentions of `FOO="1"`
+    /// // removes any mentions of `FOO="2"`
+    /// // removes any mentions of `BAR`, no matter the value
+    /// ```
+    ///
+    /// > assuming "FOO" with at least its values "1" and "2", as well as "BAR" with
+    /// > any values, are present in at least one environment.
+    ///
+    /// `to_remove` may be empty, nothing will be changed in that case.
+    ///
+    /// ## Errors
+    /// - Returns an `EnvError` if any variable or value cannot be edited
+    pub fn remove_from_environments(&mut self, to_remove: Vec<Vec<String>>) -> Result<()> {
+        if to_remove.is_empty() {
+            return Ok(());
+        }
+
+        // collect all existing envs
+        let to_remove: EnvVarList = transform_env_list(&to_remove)?;
+
+        for (var, vals) in &to_remove {
+            // var exists?
+            assert_exists(&self.environment_list, |env_file| {
+                env_file.contains_variable(var)
+            })
+            .map_err(|e| Error::EnvError {
+                reason: format!("Variable {var} cannot be edited: {e}"),
+            })?;
+
+            // vals exists?
+            for val in vals {
+                assert_exists(&self.environment_list, |env_file| {
+                    env_file.get_value(var).unwrap().contains(val)
+                })
+                .map_err(|e| Error::EnvError {
+                    reason: format!("Value {val} of {var} cannot be edited: {e}"),
+                })?;
+            }
+        }
+
+        // combine them, produces list of all env files with content
+        self.try_edit_values(&to_remove, EditMode::Remove)
+    }
+
+    /// Edit existing environment variables.
+    ///
+    /// Depending on `edit_mode` it will:
+    /// - `EditMode::Append`:
+    ///   Add all values in `to_edit` to the list of possible values.
+    /// - `EditMode::Remove`:
+    ///   Remove all values in `to_edit` from the list of possible values.
+    ///   Variables with empty value lists will then also be removed.
+    ///
+    /// Then calls on `try_assemble_all` to generate a "list of files" so to say, which
+    /// contains all possible combinations of all values.
+    /// This list will replace the current `self.environment_list`.
+    ///
+    /// Duplicate values will be removed before creating this list.
+    ///
+    /// # Panics
+    /// - panics if a key from `to_edit` cannot be found in `self.environment_list`
+    fn try_edit_values(&mut self, to_edit: &EnvVarList, edit_mode: EditMode) -> Result<()> {
```

Comment:
This should not exist. Split append and remove part and put them in their respective functions